File: scripts/amazon/check_bestseller.js
import { chromium } from 'playwright';
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

const PRODUCTS = [
  {
    name: 'GK0486-Master',
    url: 'https://www.amazon.com/dp/B08L8VZHKR/ref=twister_B07DT6KJWQ?_encoding=UTF8&th=1',
  },
  {
    name: 'PN10-Maste',
    url: 'https://www.amazon.com/dp/B004MZMBC4?th=1',
  },
  {
    name: 'PL6328-Master',
    url: 'https://www.amazon.com/dp/B0785J3B94?th=1&psc=1',
  },
];

const HISTORY_FILE = join(__dirname, 'history.json');
const REPORT_FILE = join(__dirname, 'report.html');

function loadHistory() {
  if (!existsSync(HISTORY_FILE)) return [];
  try {
    return JSON.parse(readFileSync(HISTORY_FILE, 'utf8'));
  } catch {
    return [];
  }
}

function saveHistory(history) {
  writeFileSync(HISTORY_FILE, JSON.stringify(history, null, 2));
}

async function checkBestSeller(page, product) {
  try {
    await page.goto(product.url, { waitUntil: 'domcontentloaded', timeout: 45000 });
    await page.waitForTimeout(3000);

    const result = await page.evaluate(() => {
      // Primary: dedicated badge section Amazon renders
      const badgeDiv = document.getElementById('acBadge_feature_div');
      if (badgeDiv) {
        const txt = badgeDiv.textContent || '';
        if (/best\s*seller/i.test(txt)) return { found: true, source: 'acBadge' };
      }

      // Secondary: any badge text element
      const badgeTexts = document.querySelectorAll(
        '.a-badge-text, [class*="badge-label"], [class*="bestseller"], [class*="best-seller"]'
      );
      for (const el of badgeTexts) {
        if (/best\s*seller/i.test(el.textContent || '')) {
          return { found: true, source: 'badge-text' };
        }
      }

      // Tertiary: leftNav or feature-name attribute
      const featureEl = document.querySelector('[data-feature-name="acBadge"]');
      if (featureEl && /best\s*seller/i.test(featureEl.textContent || '')) {
        return { found: true, source: 'feature-name' };
      }

      // Broad fallback — limit to above-the-fold area to avoid false positives in reviews
      const topSection = document.querySelector('#dp-container, #ppd, #centerCol');
      if (topSection) {
        const txt = topSection.textContent || '';
        if (/#1\s+best\s+seller/i.test(txt) || /best\s+seller\s+in/i.test(txt)) {
          return { found: true, source: 'text-search' };
        }
      }

      return { found: false, source: null };
    });

    return result.found;
  } catch (err) {
    console.error(`  Error on ${product.name}: ${err.message}`);
    return null;
  }
}

function buildHtml(today, results, history) {
  const statusBadge = (status) => {
    if (status === 'YES') return '<span style="color:#2e7d32;font-weight:bold">✅ YES</span>';
    if (status === 'NO') return '<span style="color:#c62828;font-weight:bold">❌ NO</span>';
    return '<span style="color:#f57f17;font-weight:bold">⚠️ ERROR</span>';
  };

  const todayRows = results.map(r => `
    <tr>
      <td><a href="${r.url}" style="color:#0645ad">${r.name}</a></td>
      <td>${today}</td>
      <td>${statusBadge(r.status)}</td>
    </tr>`).join('');

  // Last 30 days history table
  const last30 = history.slice(-30).reverse();
  const historyRows = last30.flatMap(entry =>
    entry.results.map(r => `
    <tr>
      <td><a href="${r.url}" style="color:#0645ad">${r.name}</a></td>
      <td>${entry.date}</td>
      <td>${statusBadge(r.status)}</td>
    </tr>`)
  ).join('');

  return `<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<style>
  body { font-family: Arial, sans-serif; max-width: 700px; margin: 20px auto; color: #333; }
  h2 { color: #232f3e; border-bottom: 2px solid #ff9900; padding-bottom: 8px; }
  h3 { color: #555; margin-top: 30px; }
  table { border-collapse: collapse; width: 100%; margin-top: 12px; }
  th { background: #232f3e; color: #fff; padding: 10px 14px; text-align: left; }
  td { padding: 9px 14px; border-bottom: 1px solid #eee; }
  tr:hover td { background: #f9f9f9; }
</style>
</head>
<body>
<h2>Amazon Best Seller Report — ${today}</h2>

<h3>Today's Results</h3>
<table>
  <tr><th>Product</th><th>Date</th><th>Best Seller?</th></tr>
  ${todayRows}
</table>

<h3>History (last 30 days)</h3>
<table>
  <tr><th>Product</th><th>Date</th><th>Best Seller?</th></tr>
  ${historyRows}
</table>
</body>
</html>`;
}

async function main() {
  const today = new Date().toISOString().split('T')[0];
  const history = loadHistory();
  const results = [];

  console.log(`Amazon Best Seller check — ${today}`);

  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage', '--disable-gpu'],
  });

  try {
    const context = await browser.newContext({
      userAgent:
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      viewport: { width: 1280, height: 800 },
      extraHTTPHeaders: {
        'Accept-Language': 'en-US,en;q=0.9',
        Accept:
          'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
      },
    });

    const page = await context.newPage();

    for (const product of PRODUCTS) {
      process.stdout.write(`  ${product.name} ... `);
      const hasBadge = await checkBestSeller(page, product);
      const status = hasBadge === null ? 'ERROR' : hasBadge ? 'YES' : 'NO';
      console.log(status);

      results.push({ name: product.name, url: product.url, date: today, status });

      await page.waitForTimeout(4000);
    }

    await context.close();
  } finally {
    await browser.close();
  }

  // Update history (replace today's entry if re-running)
  const updatedHistory = history.filter(e => e.date !== today);
  updatedHistory.push({ date: today, results });
  saveHistory(updatedHistory);

  writeFileSync(REPORT_FILE, buildHtml(today, results, updatedHistory));

  console.log('\nSaved: history.json, report.html');

  // Print summary for CI logs
  for (const r of results) {
    console.log(`RESULT: ${r.name} | ${r.date} | ${r.status}`);
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/check_bestseller.py
#!/usr/bin/env python3
"""Daily Amazon Best Seller badge checker.

Checks each listing for the Best Seller badge, appends to history JSON,
and sends a report email via Gmail SMTP.

Required env vars:
  GMAIL_USER          - Gmail address to send from
  GMAIL_APP_PASSWORD  - Gmail App Password (not your regular password)
  RECIPIENT_EMAIL     - Where to send the report (default: denis@bodyj4you.com)
"""

import asyncio
import json
import os
import smtplib
import sys
from datetime import datetime, timezone
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from pathlib import Path

from playwright.async_api import async_playwright

LISTINGS = [
    {
        "name": "GK0486-Master",
        "url": "https://www.amazon.com/dp/B08L8VZHKR/ref=twister_B07DT6KJWQ?_encoding=UTF8&th=1",
        "display_url": "https://www.amazon.com/dp/B08L8VZHKR",
    },
    {
        "name": "PN10-Master",
        "url": "https://www.amazon.com/dp/B004MZMBC4?th=1",
        "display_url": "https://www.amazon.com/dp/B004MZMBC4",
    },
    {
        "name": "PL6328-Master",
        "url": "https://www.amazon.com/dp/B0785J3B94?th=1&psc=1",
        "display_url": "https://www.amazon.com/dp/B0785J3B94",
    },
]

HISTORY_FILE = Path(__file__).parent.parent / "amazon_bestseller_history.json"

RECIPIENT_EMAIL = os.environ.get("RECIPIENT_EMAIL", "denis@bodyj4you.com")


async def check_best_seller(page, listing: dict) -> str:
    """Return 'YES', 'NO', or 'ERROR: ...' for a listing."""
    try:
        await page.goto(listing["url"], wait_until="domcontentloaded", timeout=30000)
        await page.wait_for_timeout(3000)

        content = await page.content()

        # Detect block / CAPTCHA pages (very short = something wrong)
        if len(content) < 5000:
            return "ERROR: page too short (possible block)"
        if "robot check" in content.lower() or "enter the characters" in content.lower():
            return "ERROR: CAPTCHA detected"

        # Primary: look for best seller badge text
        if "#1 Best Seller" in content or "Best Seller" in content:
            return "YES"

        return "NO"

    except Exception as e:
        return f"ERROR: {str(e)[:120]}"


async def run_checks() -> tuple[list[dict], str]:
    """Run all listing checks and return (results, today_date)."""
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    results = []

    async with async_playwright() as p:
        browser = await p.chromium.launch(
            headless=True,
            args=[
                "--no-sandbox",
                "--disable-setuid-sandbox",
                "--disable-dev-shm-usage",
                "--disable-gpu",
            ],
        )
        context = await browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/120.0.0.0 Safari/537.36"
            ),
            viewport={"width": 1280, "height": 900},
            extra_http_headers={
                "Accept-Language": "en-US,en;q=0.9",
                "Accept": (
                    "text/html,application/xhtml+xml,application/xml;"
                    "q=0.9,image/webp,*/*;q=0.8"
                ),
            },
        )
        page = await context.new_page()

        for listing in LISTINGS:
            status = await check_best_seller(page, listing)
            entry = {
                "date": today,
                "name": listing["name"],
                "url": listing["display_url"],
                "best_seller": status,
            }
            results.append(entry)
            print(f"  {listing['name']}: {status}")

        await browser.close()

    return results, today


def load_history() -> list[dict]:
    if HISTORY_FILE.exists():
        with open(HISTORY_FILE) as f:
            return json.load(f)
    return []


def save_history(history: list[dict]) -> None:
    with open(HISTORY_FILE, "w") as f:
        json.dump(history, f, indent=2)


def _status_cell(status: str) -> str:
    if status == "YES":
        return '<td style="color:#27ae60;font-weight:bold;text-align:center;">✅ YES</td>'
    if status == "NO":
        return '<td style="color:#e74c3c;text-align:center;">❌ NO</td>'
    return f'<td style="color:#f39c12;text-align:center;">⚠️ {status}</td>'


def build_email_html(results: list[dict], today: str, history: list[dict]) -> str:
    today_rows = ""
    for r in results:
        today_rows += f"""
        <tr>
          <td><a href="{r['url']}" style="color:#1a73e8;">{r['name']}</a></td>
          <td style="text-align:center;">{r['date']}</td>
          {_status_cell(r['best_seller'])}
        </tr>"""

    # Last 30 historical entries (excluding today, already shown above)
    past = [e for e in history if e["date"] < today][-30:]
    history_rows = ""
    for r in reversed(past):
        history_rows += f"""
        <tr>
          <td><a href="{r['url']}" style="color:#1a73e8;">{r['name']}</a></td>
          <td style="text-align:center;">{r['date']}</td>
          {_status_cell(r['best_seller'])}
        </tr>"""

    history_section = ""
    if history_rows:
        history_section = f"""
        <h3 style="margin-top:32px;">Previous Checks</h3>
        <table {TABLE_STYLE}>
          <tr style="background:#f5f5f5;">
            <th style="text-align:left;padding:8px;">Product</th>
            <th style="padding:8px;">Date</th>
            <th style="padding:8px;">Best Seller?</th>
          </tr>
          {history_rows}
        </table>"""

    return f"""
<html>
<body style="font-family:Arial,sans-serif;color:#333;max-width:640px;margin:auto;padding:20px;">
  <h2 style="border-bottom:2px solid #1a73e8;padding-bottom:8px;">
    Amazon Best Seller Daily Report — {today}
  </h2>

  <h3>Today's Results</h3>
  <table {TABLE_STYLE}>
    <tr style="background:#f5f5f5;">
      <th style="text-align:left;padding:8px;">Product</th>
      <th style="padding:8px;">Date</th>
      <th style="padding:8px;">Best Seller?</th>
    </tr>
    {today_rows}
  </table>
  {history_section}

  <p style="color:#999;font-size:12px;margin-top:32px;">
    Automated check · History stored in
    <a href="https://github.com/frolovdo/bodyj4you" style="color:#999;">
      frolovdo/bodyj4you
    </a>
  </p>
</body>
</html>"""


TABLE_STYLE = (
    'border="1" cellpadding="8" cellspacing="0" '
    'style="border-collapse:collapse;width:100%;"'
)


def send_email(results: list[dict], today: str, history: list[dict]) -> None:
    gmail_user = os.environ.get("GMAIL_USER")
    gmail_password = os.environ.get("GMAIL_APP_PASSWORD")

    if not gmail_user or not gmail_password:
        print("WARNING: GMAIL_USER / GMAIL_APP_PASSWORD not set — skipping email.")
        return

    html_body = build_email_html(results, today, history)

    msg = MIMEMultipart("alternative")
    msg["Subject"] = f"Amazon Best Seller Check — {today}"
    msg["From"] = gmail_user
    msg["To"] = RECIPIENT_EMAIL
    msg.attach(MIMEText(html_body, "html"))

    with smtplib.SMTP_SSL("smtp.gmail.com", 465) as server:
        server.login(gmail_user, gmail_password)
        server.send_message(msg)

    print(f"Email sent to {RECIPIENT_EMAIL}")


def main() -> None:
    print(f"Checking Amazon listings...")
    results, today = asyncio.run(run_checks())

    history = load_history()
    history.extend(results)
    save_history(history)
    print(f"History updated ({len(history)} total entries)")

    send_email(results, today, history)

    print("\nDone.")
    for r in results:
        print(f"  {r['name']}: {r['best_seller']}")


if __name__ == "__main__":
    main()
